### ISSUE TITLE
Singularity and docker

### ISSUE THREAD

**Issue by photocyte** [blobtoolkit/viewer#5]
Hi there,

I'm trying to view an `insdc-pipeline` produced blobdir, via the blobtoolkit Docker image being run via singularity on a Linux cluster node.  The image is pulled and executed as follows:
```
singularity pull docker://genomehubs/blobtoolkit
export VIEWER=true; singularity run -B ../insdc-pipeline/btk_workdir/dir_for_blobdirs/:/blobtoolkit/datasets blobtoolkit_latest.sif
```

Though `singularity run`, `node` seems to boot successfully, and after ssh port forwarding e.g. `8080:127.0.0.1:8080 -L 8000:127.0.0.1:8000`, I'm getting listening services on both 8000 and 8080. I can confirm I can see the blobdir in the `/blobtoolkit/datasets` folder.
```
singularity shell -B ../insdc-pipeline/btk_workdir/dir_for_blobdirs/:/blobtoolkit/datasets blobtoolkit_latest.sif
Singularity> ls /blobtoolkit/datasets
Pnm11bph1
Singularity> ls /blobtoolkit/datasets/Pnm11bph1/
L319l1_cov.json		  bestsumorder_class.json	      bestsumorder_genus_score.json	   bestsumorder_phylum_positions.json	     endopterygota_odb9_busco.json
L319l1_read_cov.json	  bestsumorder_class_cindex.json      bestsumorder_kingdom.json		   bestsumorder_phylum_score.json	     eukaryota_odb9_busco.json
L319l2_cov.json		  bestsumorder_class_positions.json   bestsumorder_kingdom_cindex.json	   bestsumorder_positions.json		     gc.json
L319l2_read_cov.json	  bestsumorder_class_score.json       bestsumorder_kingdom_positions.json  bestsumorder_species.json		     identifiers.json
L319l3_cov.json		  bestsumorder_family.json	      bestsumorder_kingdom_score.json	   bestsumorder_species_cindex.json	     length.json
L319l3_read_cov.json	  bestsumorder_family_cindex.json     bestsumorder_order.json		   bestsumorder_species_positions.json	     meta.json
L319l4_cov.json		  bestsumorder_family_positions.json  bestsumorder_order_cindex.json	   bestsumorder_species_score.json	     ncount.json
L319l4_read_cov.json	  bestsumorder_family_score.json      bestsumorder_order_positions.json    bestsumorder_superkingdom.json
L319l5_cov.json		  bestsumorder_genus.json	      bestsumorder_order_score.json	   bestsumorder_superkingdom_cindex.json
L319l5_read_cov.json	  bestsumorder_genus_cindex.json      bestsumorder_phylum.json		   bestsumorder_superkingdom_positions.json
bacteria_odb9_busco.json  bestsumorder_genus_positions.json   bestsumorder_phylum_cindex.json	   bestsumorder_superkingdom_score.json
```

If I do a http request to `127.0.0.1:8000/`, I get:
`Cannot GET /`

If I do a a http request to `127.0.0.1:8080/view/all`, I see this splash screen:
<img width="477" alt="Screen Shot 2020-04-29 at 11 24 40 AM" src="https://user-images.githubusercontent.com/12517705/80632697-78412f80-8a0c-11ea-911a-040f84425baf.png">
, but then end up w/ a blank white screen within 5 seconds or so.  Any thoughts on what might be going wrong?

**Comment by photocyte** [comment]
It seems if I use `http://localhost:8080/view/Pnm11bph1/` instead of `http://127.0.0.1:8080/view/Pnm11bph1/`, it does work.  I can find my dataset if searching in the "Search for datasets..." textbox, but, if I click on it, it just shows the "Loading dataset" splash briefly, then goes back to the main page.  I assume something might be a bit malformed about my BlobDir, so I'll look into that...

**Comment by photocyte** [comment]
So, I've spent a bit more time exploring. Part of the issue was I didn't know what to expect from a working Blobtoolkit viewer instance, but I don't think that is the whole problem. 

For example, if we look at one of your pre-computed analyses online, there are several tabs that produce informative plots right from the "Filter" interface:
https://blobtoolkit.genomehubs.org/view/Daphnia%20pulex/dataset/ACJG01/blob#Filters
<img width="436" alt="Screen Shot 2020-04-29 at 12 27 36 PM" src="https://user-images.githubusercontent.com/12517705/80638553-7465db00-8a15-11ea-8656-e511902c44c1.png">

In my case, (viewing my `insdc-pipeline` produced BlobDir using the ssh forwarded Singularity container hosted blobtoolkit/viewer at `localhost:8080`), those tabs are missing:
<img width="446" alt="Screen Shot 2020-04-29 at 12 27 45 PM" src="https://user-images.githubusercontent.com/12517705/80638580-847dba80-8a15-11ea-8fa2-1c883ace003e.png">

It does seem that some of the data files were successfully parsed / loaded, like the coverage files, but hitting the X vs Y vs Z axes buttons doesn't change anything (nothing being plotted on right - just a blank white space):
<img width="397" alt="Screen Shot 2020-04-29 at 12 33 22 PM" src="https://user-images.githubusercontent.com/12517705/80638732-c27ade80-8a15-11ea-8360-15aa22302727.png">

If I go into the "Settings" interface, I'm not getting the blob plot, busco plot, etc.  Snail plot does work though 
<img width="880" alt="Screen Shot 2020-04-29 at 12 33 35 PM" src="https://user-images.githubusercontent.com/12517705/80638774-d1619100-8a15-11ea-83c3-530ff7b4e5c6.png">

Still suspecting that something is strange about the BlobDir.  Since I edited the `indsc-pipeline` config file to fix some errors (described here: https://github.com/blobtoolkit/insdc-pipeline/issues/1), while the pipeline was in an intermediate state, perhaps that is the culprit, so I'll `snakemake --delete-all-output` and try to rerun pipeline from scratch. Let me know if it would help to send the BlobDir over to you to see what might be going on?

**Comment by rjchallis** [comment]
Hi

To explain the problem that you already solved, the get request to `127.0.0.1:8000` is rejected by the API running as it only allows requests from localhost by default, this can be configured by updating `BTK_ORIGINS` in the `.env` file to include `127.0.0.1` and `127.0.0.1:8080` (or exporting it directly as an environment variable). I can update the default to include `127.0.0.1` to fix this. I'm also planning to add more informative messages to the viewer, rather than leaving it to crash or hang, which isn't very user friendly at the moment.

The missing plot types suggest something awry with the BlobDir as you suggest. You could try running the validator (see [blobtoolkit.genomehubs.org/specification/specification-tutorials/validating-datasets/](https://blobtoolkit.genomehubs.org/specification/specification-tutorials/validating-datasets/)) which will test for most errors in the BlobDir. If you have x, y and z active but nothing is plotting then it looks like the categories may be the problem (they aren't used by the snail plot). If you scroll down the filters do you have a preview plot for `bestsumorder_phylum` and is the category axis icon active?

**Comment by photocyte** [comment]
Hi there,  still having some issues using the viewer locally w/ Singularity, using the up-to-date but admittedly converted Docker container from here:
`singularity pull docker://genomehubs/blobtoolkit:2.6.5`

I'm trying to view the `BDQP01` example Blobdir, which I got from `/blobtoolkit/viewer/demo/BDQP01` inside the container.
```
singularity exec blobtoolkit_2.6.5.sif ls /blobtoolkit/viewer/demo/BDQP01
DRR055278_cov.json       ILLUMINA.sum_cov.json            bestsumorder_genus_cindex.json  bestsumorder_phylum_cindex.json   bestsumorder_superkingdom_cindex.json  ncount.json
DRR055278_read_cov.json  ILLUMINA.sum_read_cov.json       bestsumorder_genus_score.json   bestsumorder_phylum_score.json    bestsumorder_superkingdom_score.json
DRR055279_cov.json       bestsumorder_family.json         bestsumorder_order.json         bestsumorder_species.json         gc.json
DRR055279_read_cov.json  bestsumorder_family_cindex.json  bestsumorder_order_cindex.json  bestsumorder_species_cindex.json  identifiers.json
DRR055280_cov.json       bestsumorder_family_score.json   bestsumorder_order_score.json   bestsumorder_species_score.json   length.json
DRR055280_read_cov.json  bestsumorder_genus.json          bestsumorder_phylum.json        bestsumorder_superkingdom.json    meta.json
```

I moved the example Blobdir out of the container, and can successfully mount it the parent directory & it back into the container:
```
singularity exec -B ./blobdirs:/blobtoolkit/datasets blobtoolkit_2.6.5.sif ls /blobtoolkit/datasets/BDQP01
DRR055278_cov.json       ILLUMINA.sum_cov.json            bestsumorder_genus_cindex.json  bestsumorder_phylum_cindex.json   bestsumorder_superkingdom_cindex.json  ncount.json
DRR055278_read_cov.json  ILLUMINA.sum_read_cov.json       bestsumorder_genus_score.json   bestsumorder_phylum_score.json    bestsumorder_superkingdom_score.json
DRR055279_cov.json       bestsumorder_family.json         bestsumorder_order.json         bestsumorder_species.json         gc.json
DRR055279_read_cov.json  bestsumorder_family_cindex.json  bestsumorder_order_cindex.json  bestsumorder_species_cindex.json  identifiers.json
DRR055280_cov.json       bestsumorder_family_score.json   bestsumorder_order_score.json   bestsumorder_species_score.json   length.json
DRR055280_read_cov.json  bestsumorder_genus.json          bestsumorder_phylum.json        bestsumorder_superkingdom.json    meta.json
```

Notably, this example Blobdir fails some validations using `validate.py`, but I'll assume it's okay?
```
singularity exec -B ./blobdirs:/blobtoolkit/datasets blobtoolkit_2.6.5.sif validate.py /blobtoolkit/datasets/BDQP01
NOT VALID: Error while testing raw metadata against schema
data.paired must be object
```

In trying to use `blobtools host` I seemingly get a usable BlobToolKit Viewer instance. (real hostname redacted)
```
singularity exec -B ./blobdirs:/blobtoolkit/datasets blobtoolkit_2.6.5.sif blobtools host --hostname [HOSTNAME REDACTED] /blobtoolkit/datasets/BDQP01
WARNING: Directory '/blobtoolkit/datasets/BDQP01' appears to be a BlobDir.
         Hosting the parent directory instead.
Starting BlobToolKit API on port 8000 (pid: 213630)
Starting BlobToolKit viewer on port 8080 (pid: 213651)
Visit http://[HOSTNAME REDACTED]:8080 to use the interactive BlobToolKit Viewer.
```

But, when I navigate my browser to `http://[HOSTNAME REDACTED]:8080`, no dataset(s) shows up. This is especially perplexing, as the `WARNING` above seems to suggest `blobtools host` was able to partially parse the Blobdir - so I assume it isn't a permissions error or file availability error.
<img width="908" alt="Screen Shot 2022-01-21 at 2 15 45 PM" src="https://user-images.githubusercontent.com/12517705/150616281-e6394458-5714-4474-ac41-566225feb038.png">

I'll note, I also tried with the `FXWY01` Blobdir from `/blobtoolkit/specification/example/FXWY01`, that one does pass the validation, but it alone or with `BDQP01` in the mounted `blobdirs` directory used in the example code blocks above, have the same overall issue where `FXWY01` and/or `BDQP01` do not show up in the datasets tab of the BlobToolKit Viewer. Do you have any tips on what might be going wrong?
